File: web/src/test/java/no/nav/modiapersonoversikt/service/unleash/strategier/ByEnhetStrategyTest.java
package no.nav.modiapersonoversikt.service.unleash.strategier;

import no.finn.unleash.UnleashContext;
import org.junit.jupiter.api.Test;

import java.util.HashMap;

import static no.nav.modiapersonoversikt.service.unleash.strategier.ByEnhetStrategy.ENHETER;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class ByEnhetStrategyTest {

    private final ByEnhetStrategy byEnhetStrategy = new ByEnhetStrategy();
    private final UnleashContext context = mock(UnleashContext.class);

    @Test
    void enhetTest() {
        assertEnabled("0118", "0118");
        assertEnabled("0118", "1234,0118,0000");
        assertEnabled("0118,1234,1111", "1234");
        assertEnabled("0118,1234,1111", "0118,1234,1111");
        assertDisabled("0118", "1234, 4455");
        assertDisabled("0118", "");
        assertDisabled("0118", null);
        assertDisabled("", "1234");
        assertDisabled(null, "1234");
        assertDisabled("", "");
        assertDisabled(null, null);
        assertDisabled("", ",,,");
        assertDisabled(",,,", ",,,");
    }

    @Test
    void kallMedIngenContextGirFalse() {
        HashMap<String, String> parameters = new HashMap<String, String>() {{
            put(ENHETER, "0118");
        }};

        assertThat(byEnhetStrategy.isEnabled(parameters), is(false));
    }

    private void assertEnabled(String enheter, String parameter) {
        addEnhetToContextProperties(enheter);

        assertEnhet(parameter, true);
    }

    private void assertDisabled(String enheter, String parameter) {
        addEnhetToContextProperties(enheter);

        assertEnhet(parameter, false);
    }

    private void addEnhetToContextProperties(String enheter) {
        HashMap<String, String> props = new HashMap<String, String>() {{
            put(ENHETER, enheter);
        }};
        when(context.getProperties()).thenReturn(props);
    }

    private void assertEnhet(String parameter, boolean expected) {
        HashMap<String, String> params = new HashMap<String, String>() {{
            put("valgtEnhet", parameter);
        }};
        assertThat(byEnhetStrategy.isEnabled(params, context), is(expected));
    }
}

File: web/src/test/java/no/nav/modiapersonoversikt/service/unleash/strategier/ByEnhetStrategyTest.kt
package no.nav.modiapersonoversikt.service.unleash.strategier

import io.mockk.every
import io.mockk.mockk
import no.finn.unleash.UnleashContext
import no.nav.modiapersonoversikt.service.unleash.strategier.StrategyUtils.ENHETER_PROPERTY
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

internal class ByEnhetStrategyTest {
    val byEnhetStrategy = ByEnhetStrategy()
    val unleashContext = mockk<UnleashContext>()

    @Test
    internal fun `calls without context should return false`() {
        assertThat(byEnhetStrategy.isEnabled(emptyMap())).isFalse
    }

    @ParameterizedTest
    @MethodSource("testcases")
    internal fun `should evaluate strategy`(unleashConfig: String?, ansattesEnheter: String?, expected: Boolean) {
        val parameters = mapOf(ByEnhetStrategy.ENABLED_ENHETER_PROPERTY to unleashConfig)
        every { unleashContext.properties } returns mapOf(ENHETER_PROPERTY to ansattesEnheter)

        assertThat(byEnhetStrategy.isEnabled(parameters, unleashContext)).isEqualTo(expected)
    }

    companion object {
        @JvmStatic
        fun testcases() = listOf(
            strategyEnabled(unleashConfig = "0118", ansattesEnheter = "0118"),
            strategyEnabled(unleashConfig = "0118", ansattesEnheter = "1234,0118,0000"),
            strategyEnabled(unleashConfig = "0118,1234,1111", ansattesEnheter = "1234"),
            strategyEnabled(unleashConfig = "0118,1234,1111", ansattesEnheter = "0118,1234,1111"),
            strategyDisabled(unleashConfig = "0118", ansattesEnheter = "1234, 4455"),
            strategyDisabled(unleashConfig = "0118", ansattesEnheter = ""),
            strategyDisabled(unleashConfig = "0118", ansattesEnheter = null),
            strategyDisabled(unleashConfig = "", ansattesEnheter = "1234"),
            strategyDisabled(unleashConfig = null, ansattesEnheter = "1234"),
            strategyDisabled(unleashConfig = "", ansattesEnheter = ""),
            strategyDisabled(unleashConfig = null, ansattesEnheter = null),
            strategyDisabled(unleashConfig = "", ansattesEnheter = ",,,"),
            strategyDisabled(unleashConfig = ",,,", ansattesEnheter = ",,,"),
        )

        private fun strategyEnabled(unleashConfig: String?, ansattesEnheter: String?): Arguments {
            return Arguments.of(unleashConfig, ansattesEnheter, true)
        }
        private fun strategyDisabled(unleashConfig: String?, ansattesEnheter: String?): Arguments {
            return Arguments.of(unleashConfig, ansattesEnheter, false)
        }
    }
}
